File: app/src/main/java/www/sanju/saviour/MainActivity.java
package www.sanju.saviour;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.androidnetworking.AndroidNetworking;
import com.androidnetworking.common.Priority;
import com.androidnetworking.error.ANError;
import com.androidnetworking.interfaces.StringRequestListener;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.Circle;
import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import org.json.JSONException;
import org.json.JSONObject;

public class MainActivity extends AppCompatActivity implements OnMapReadyCallback {
    private GoogleMap mMap;
    private Circle mCircle;
    public static final String USER_NAME = "Karthick";


    private static final String URL = "https://api.thingspeak.com/channels/1009199/feeds/last.json?api_key=VRMW2SVE9KK9YD50";
    private static final String TAG = "MainActivity";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        AndroidNetworking.initialize(getApplicationContext());

        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map_main);
        mapFragment.getMapAsync(this);

        fetchLatestAccident();
    }



    private void fetchLatestAccident() {
        AndroidNetworking.get(URL)
                .setPriority(Priority.HIGH)
                .build()
                .getAsString(new StringRequestListener() {
                    @Override
                    public void onResponse(String response) {
                        Log.i(TAG, "onResponse: " + response);

                        if (response.length() > 0) {

                            try {
                                JSONObject result = new JSONObject(response);

                                String lat = result.getString("field1");
                                String lon = result.getString("field2");
                                int userid = result.getInt("entry_id");
                                String id = String.valueOf(userid);


                                double latitude=Double.parseDouble(lat);
                                double longitude=Double.parseDouble(lon);

                                showLocation(latitude, longitude, id);





                            } catch (JSONException e) {
                                e.printStackTrace();
                            }


                           // Toast.makeText(MainActivity.this, "Data fetched" + response, Toast.LENGTH_SHORT).show();

                        } else {

                            Toast.makeText(MainActivity.this, "No Accident Detected", Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onError(ANError error) {
                        Toast.makeText(MainActivity.this, " " + error, Toast.LENGTH_SHORT).show();
                    }
                });

    }

    @Override
    public void onMapReady(GoogleMap googleMap) {

        mMap = googleMap;

    }

    private void showLocation(double latitude, double longitude, String id) {

        LatLng accidentLocation = new LatLng(latitude, longitude);

//        // circle option & custom marker
                         mCircle = mMap.addCircle(new CircleOptions()
                                .center(accidentLocation)
                                .radius(50)
                                .strokeColor(getResources().getColor(R.color.accident_zone))
                                .fillColor(getResources().getColor(R.color.accident_zone_trans)));

                        mMap.addMarker(new MarkerOptions().position(accidentLocation).title(id.concat(" ").concat(USER_NAME))
                                .icon(bitmapDescriptorFromVector(MainActivity.this, R.drawable.ic_warning_black_24dp)));
        mMap.setMyLocationEnabled(true);
        mMap.animateCamera(CameraUpdateFactory.newLatLngZoom(accidentLocation, 18f));
    }


    // custom marker for saviour
    private BitmapDescriptor bitmapDescriptorFromVector(Context context, int vectorResId) {
        Drawable vectorDrawable = ContextCompat.getDrawable(context, vectorResId);
        vectorDrawable.setBounds(0, 0, vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight());
        Bitmap bitmap = Bitmap.createBitmap(vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        vectorDrawable.draw(canvas);
        return BitmapDescriptorFactory.fromBitmap(bitmap);
    }


}

File: app/src/main/java/www/sanju/saviour/MainActivity.kt
package www.sanju.saviour

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy.KEEP
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import com.androidnetworking.AndroidNetworking
import com.androidnetworking.common.Priority
import com.androidnetworking.error.ANError
import com.androidnetworking.interfaces.StringRequestListener
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import org.json.JSONException
import org.json.JSONObject
import java.util.concurrent.TimeUnit


class MainActivity : AppCompatActivity(), OnMapReadyCallback {
    private var mMap: GoogleMap? = null
    private var mCircle: Circle? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        AndroidNetworking.initialize(applicationContext)

        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        val mapFragment = supportFragmentManager
                .findFragmentById(R.id.map_main) as SupportMapFragment?
        mapFragment!!.getMapAsync(this)

        // get accidents from thinkSpeak
        fetchLatestAccident()
        initWorker()
    }


    private fun initWorker() {

        // worker constraints
        val constraints = Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .setRequiresBatteryNotLow(true)
                .build()


        // periodic worker for every 15 minutes
        val notificationWorkRequest: PeriodicWorkRequest = PeriodicWorkRequest.Builder(MyWorker::class.java, 15,
                TimeUnit.MINUTES)
                .setConstraints(constraints)
                .build()


        // enqueue work
        val workManager = WorkManager.getInstance(applicationContext)
        workManager.enqueueUniquePeriodicWork(JOB_TAG, KEEP, notificationWorkRequest)

    }

    private fun fetchLatestAccident() {
        AndroidNetworking.get(URL)
                .setPriority(Priority.HIGH)
                .build()
                .getAsString(object : StringRequestListener {
                    override fun onResponse(response: String) {
                        Log.i(TAG, "onResponse: $response")
                        if (response.isNotEmpty()) {
                            try {
                                val result = JSONObject(response)
                                val lat = result.getString("field1")
                                val lon = result.getString("field2")
                                val userid = result.getInt("entry_id")
                                val id = userid.toString()
                                val latitude = lat.toDouble()
                                val longitude = lon.toDouble()
                                showLocation(latitude, longitude, id)
                            } catch (e: JSONException) {
                                e.printStackTrace()
                            }

                        } else {
                            Toast.makeText(this@MainActivity, "No Accident Detected", Toast.LENGTH_SHORT).show()
                        }
                    }

                    override fun onError(error: ANError) {
                        Toast.makeText(this@MainActivity, " $error", Toast.LENGTH_SHORT).show()
                    }
                })
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
    }

    private fun showLocation(latitude: Double, longitude: Double, id: String) {
        val accidentLocation = LatLng(latitude, longitude)

        // circle option & custom marker
        mCircle = mMap!!.addCircle(CircleOptions()
                .center(accidentLocation)
                .radius(50.0)
                .strokeColor(resources.getColor(R.color.accident_zone))
                .fillColor(resources.getColor(R.color.accident_zone_trans)))
        mMap!!.addMarker(MarkerOptions().position(accidentLocation).title(id + " " + USER_NAME)
                .icon(bitmapDescriptorFromVector(this@MainActivity, R.drawable.ic_warning_black_24dp)))
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            // TODO: Consider calling
            //    ActivityCompat#requestPermissions
            // here to request the missing permissions, and then overriding

            return
        }
        mMap!!.isMyLocationEnabled = true
        mMap!!.animateCamera(CameraUpdateFactory.newLatLngZoom(accidentLocation, 18f))
    }

    // custom marker for saviour
    private fun bitmapDescriptorFromVector(context: Context, vectorResId: Int): BitmapDescriptor {
        val vectorDrawable = ContextCompat.getDrawable(context, vectorResId)
        vectorDrawable!!.setBounds(0, 0, vectorDrawable.intrinsicWidth, vectorDrawable.intrinsicHeight)
        val bitmap = Bitmap.createBitmap(vectorDrawable.intrinsicWidth, vectorDrawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        vectorDrawable.draw(canvas)
        return BitmapDescriptorFactory.fromBitmap(bitmap)
    }

    companion object {
        const val USER_NAME = "Karthick"
        private const val URL = "https://api.thingspeak.com/channels/1009199/feeds/last.json?api_key=VRMW2SVE9KK9YD50"
        private const val TAG = "MainActivity"
        private const val JOB_TAG = "AccidentWorker"
    }
}